docs/mypy_plugin.md

```diff
@@ -0,0 +1,137 @@
+Pydantic works well with [mypy](http://mypy-lang.org/) right [out of the box](usage/mypy.md).
+
+However, Pydantic also ships with a mypy plugin that adds a number of important pydantic-specific
+features to mypy that improve its ability to type-check your code.
+
+For example, consider the following script:
+```py
+{!.tmp_examples/mypy.py!}
+```
+
+Without any set up, mypy catches one of the errors:
+```
+13: error: "Model" has no attribute "middle_name"
+```
+
+But with the plugin enabled, it catches both:
+```
+13: error: "Model" has no attribute "middle_name"
+16: error: Missing named argument "age" for "Model"
+16: error: Missing named argument "list_of_ints" for "Model"
+```
+
+With the pydantic mypy plugin, you can fearlessly refactor your models knowing mypy will catch any mistakes
+if your field names or types change.
+
+There are other benefits too! See below for more details.
+
+### Plugin Capabilities
+
+#### Generate a signature for `Model.__init__`
+* Any required fields that don't have dynamically-determined aliases will be included as required
+  keyword arguments.
+* If `Config.allow_population_by_field_name=True`, the generated signature will use the field names,
+  rather than aliases.
+* For subclasses of `BaseSettings`, all fields are treated as optional since they may be read from the environment.
+* If `Config.extra="forbid"` and you don't make use of dynamically-determined aliases, the generated signature
+  will not allow unexpected inputs.
+* **Optional:** If the `init_allow_extra` **plugin setting** (see below) is set to `False`, unexpected inputs to
+  `__init__` will raise errors even if `Config.extra` is not `"forbid"`.
+* **Optional:** If the `init_typed` **plugin setting** (see below) is set to `True`, the generated signature
+  will use the types of the model fields (otherwise they will be annotated as `Any` to allow parsing).
```

Comment:
Sounds good! Let me know if there are features the pycharm plugin handles that are conspicuously missing here and I can try to add them. (E.g., validation of the signature of `validator`-decorated methods.) I think it makes sense to release before getting *everything* covered (as long as there aren't obvious bugs in provided functionality), but it would be good to track.

Reply:
I have read all your mypy code.
Does show this plugin an error when access fields on a class method? 
eg: 
```python
class Model(BaseModel):
  name: str = 'abc'
  @classmethod
  def get_default_name(cls):
       return cls.name   # <- here
```
Otherwise, It looks good :)

Reply:
That’s an interesting catch!

I doubt it shows an error, but at the same time, I’m not sure mypy would typically show an error there, eg if using a dataclass. But I agree it could be a useful error to catch, especially for newcomers to pydantic (and especially in validators).

I’m not sure how I’d override that behavior but I’ll think about it. Any insight would be appreciated though!

Either way I don’t think this needs to block a first release.